xdsl/transforms/printf_to_putchar.py

```diff
@@ -0,0 +1,243 @@
+from xdsl.builder import ImplicitBuilder
+from xdsl.dialects import arith, func, scf
+from xdsl.dialects.builtin import IndexType, IntegerType, ModuleOp, i32
+from xdsl.dialects.experimental import math
+from xdsl.dialects.printf import PrintCharOp, PrintIntOp
+from xdsl.ir import Block, MLContext, Operation, OpResult, SSAValue
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.traits import SymbolTable
+
+i8 = IntegerType(8)
+
+
+class LowerPrintCharToPutchar(RewritePattern):
+    """
+    Rewrite Pattern that rewrites printf.print_char to an
+    external function call
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: PrintCharOp, rewriter: PatternRewriter, /):
+        func_call = func.Call("putchar", [op.char], [i32])
+        cast_op = arith.ExtUIOp(op.char, i32)
+        func_call = func.Call("putchar", [cast_op], [i32])
+        # Add empty new_results, since a result is necessary for linking
+        # putchar, but the result does not exist anywhere.
+        rewriter.replace_matched_op([cast_op, func_call], new_results=[])
+
+
+class ConvertPrintIntToItoa(RewritePattern):
+    """
+    Rewrite Pattern that rewrites printf.print_int to an
+    mlir_itoa function
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: PrintIntOp, rewriter: PatternRewriter, /):
+        func_call = func.Call("mlir_itoa", [op.int], [])
+        rewriter.replace_matched_op(func_call)
+
+
+def get_mlir_itoa():
+    """
+    This function returns an MLIR func.func that holds a routine that calls
+    an external putchar implementation several times on all individual digits
+    that make up the integer.
+
+    Roughly it exists of the following steps:
+        1. If the number is zero, print zero and exit
+        2. Check if the number is negative: if so, print the minus sign
+        3. Get the number of digits for the absolute value
+        4. Go through the digits in reverse order and print them one by one
+
+    Several things to note about this specfic implementation:
+        * It does not require floating point arithmetic
+        * It does not require a printf implementation, only putchar.
+    """
+
+    # get_number_of_digits implementation
+    def get_number_of_digits(absolute_value: Operation) -> OpResult:
```

Comment:
```suggestion
    def get_number_of_digits(absolute_value: SSAValue) -> OpResult:
```
Can you please pass an SSAValue here instead? What is the advantage of Operation?

Reply:
This was not a conscious decision :)